test/Stateless.Tests/SynchronizationContextFixture.cs

```diff
@@ -0,0 +1,342 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+using Xunit.Sdk;
+
+namespace Stateless.Tests;
+
+public class SynchronizationContextFixture
+{
+    // Define a custom SynchronizationContext. All calls made to delegates should be with this context.
+    private readonly MaxConcurrencySyncContext _customSynchronizationContext = new(3);
+    private readonly List<SynchronizationContext> _capturedSyncContext = new();
+    
+    private StateMachine<State, Trigger> GetSut(State initialState = State.A)
+    {
+        return new StateMachine<State, Trigger>(initialState, FiringMode.Queued)
+        {
+            RetainSynchronizationContext = true
+        };
+    }
+    
+    private void SetSyncContext()
+    {
+        SynchronizationContext.SetSynchronizationContext(_customSynchronizationContext);
+    }
+
+    /// <summary>
+    /// Simulate a call that loses the synchronization context
+    /// </summary>
+    private async Task CaptureThenLoseSyncContext()
+    {
+        CaptureSyncContext();
+        await LoseSyncContext().ConfigureAwait(false); // ConfigureAwait false here to ensure we continue using the sync context returned by LoseSyncContext 
+    }
+    
+    private void CaptureSyncContext()
+    {
+        _capturedSyncContext.Add(SynchronizationContext.Current);
+    }
+
+    private async Task LoseSyncContext()
+    {
+        await Task.Delay(TimeSpan.FromMilliseconds(1)).ConfigureAwait(false); // Switch synchronization context and continue
```

Comment:
Note the `await Task.Delay(TimeSpan.FromMilliseconds(1))` is not guaranteed to run asynchronously and can, in fact, complete synchronously. In this case, `ConfigureAwait(false)` won't switch the synchronization context. To guarantee the asynchronous execution, use `Task.Yield()` instead.

Reply:
Fascinating! Are you able to provide an example of when this would do this - or is it simply system dependant?

Reply:
This is just an implementation detail of async/await. Any awaited code can finish synchronously (this is documented somewhere, I'll try to find the link).

Reply:
Sorry, I can't find the documentation. But I will try to prove it with the source code. Look into the `Task.Delay` implementation. It [returns](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs#L5636-L5640) `DelayPromise` instance. Its _constructor_ has the following code with a comment
```c#
if (IsCompleted)
{
    // Handle rare race condition where the timer fires prior to our having stored it into the field, in which case
    // the timer won't have been cleaned up appropriately.  This call to close might race with the Cleanup call to Close,
    // but Close is thread-safe and will be a nop if it's already been closed.
    _timer.Dispose();
}
```
It means the task can complete before it even was constructed. It means you are going to `await` the completed task. `await`'ing task is essentially calling `GetAwaiter().GetResult()`. Check the `TaskAwaiter.GetResult()` [implementation](https://github.com/microsoft/referencesource/blob/master/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs#L111-L130)
```c#
public void GetResult()
{
    ValidateEnd(m_task);
}

internal static void ValidateEnd(Task task)
{
    // Fast checks that can be inlined.
    if (task.IsWaitNotificationEnabledOrNotRanToCompletion)
    {
        // If either the end await bit is set or we're not completed successfully,
        // fall back to the slower path.
        HandleNonSuccessAndDebuggerNotification(task);
    }
}
```
And `IsWaitNotificationEnabledOrNotRanToCompletion` [implementation](https://github.com/microsoft/referencesource/blob/master/mscorlib/system/threading/Tasks/Task.cs#L913-L921)
```c#
internal bool IsWaitNotificationEnabledOrNotRanToCompletion
{
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    get
    {
        return (m_stateFlags & (Task.TASK_STATE_WAIT_COMPLETION_NOTIFICATION | Task.TASK_STATE_RAN_TO_COMPLETION))
                != Task.TASK_STATE_RAN_TO_COMPLETION;
    }
}
```
Since the task state is `TASK_STATE_RAN_TO_COMPLETION`, the `IsWaitNotificationEnabledOrNotRanToCompletion` method will return `false`. `ValidateEnd` and `GetResult` will return immediately on the same thread, and the context won't switch.

The `Task.Delay` is just an example. Any fast enough async method can complete before you `await` it, and `GetResut()` will return synchronously.

Reply:
> await'ing task is essentially calling GetAwaiter().GetResult()

It's a rude simplification. In fact, there is an [async state machine](https://sharplab.io/#v2:D4AQTAjAsAUCAMACEECsBuWsQGZlkQGFYBvWRC5PEADmQDZEBZACgEpzKyZLfkBOBgDoAIgFMANgEMAniwAqASwC2YgMoAHKQDshAMQBOAe2VNFEiYoDOYgMZHtAEyssIbNph6UAvrG9A===)
```c#
private void MoveNext()
{
    int num = <>1__state;
    try
    {
        TaskAwaiter awaiter;
        if (num != 0) // num = -1 on the first execution
        {
            awaiter = Task.Delay(TimeSpan.FromMilliseconds(1)).GetAwaiter();
            if (!awaiter.IsCompleted)
            {
                ...
                return;
            }
        }
        else
        {
            ...
        }
        awaiter.GetResult();
    }
    catch (Exception exception)
    {
        ...
    }
}
```
Since the `awaiter.IsCompleted` is `true` we are jumping straight to the `awaiter.GetResult()`.

Reply:
Thank you for taking the time to explain it! Well, I've been learning so much recently! I completely agree- I'll stick a Task.Yield in there instead.